### PR TITLE
[TT-9493] - Add Regex functions

### PR DESCRIPTION
**Pull request by chaddgrimm**
This change adds the following function to the `Geocoder` class:

- isPossible3wa - determines if the string passed in is the form of a three word address, returns 1 if valid, 0 if not
- findPossible3wa - searches the string passed in for all substrings in the form of a three word address, does not validate whether it is a real address, returns an array of possible what3word addresses.
- isValid3wa - determines of the string passed in is a real three word address, calls the API to verify it refers to an actual place on earth, returns 1 if valid, 0 if not

Also added a docker compose config for spinning up a container that can host a local php instance (more info on the readme).

### CODE REVIEW

File: src/Geocoder.php [hunk 1]
```diff
@@ -8,216 +8,252 @@
...
-            $this->error["code"]    = "BadConnection";
-            $this->error["message"] = curl_error($handle);
-            }
-
-        curl_close($handle);
-
-        // return the json
-        return $output;
-        }
+class Geocoder
```
Comment (downeyfe) [inline src/Geocoder.php:18]: Can you direct me to what's really changed here? Looks like some formatting changes are causing a lot to be flagged as new

Comment (chaddgrimm) [reply src/Geocoder.php:18]: Yeah, Github is funny when displaying the diff for this file, not sure why 🤷‍♂️. The diff looks much better in VSCode (+ GitLens). But the main change here would be the addition of the following methods:

```
  public function isPossible3wa($input);
  public function findPossible3wa($input);
  public function isValid3wa($input);
```

Also renamed a couple of variables to pascalCase, as well as added access specifiers for its properties.

File: src/Geocoder.php [hunk 2]
```diff
@@ -8,216 +8,252 @@
...
+  {
+    $regex = "/[^0-9`~!@#$%^&*()+\-_=[{\]}\|'<,.>?\/\";:£§º©®\s]+[.｡。･・︒។։။۔።।][^0-9`~!@#$%^&*()+\-_=[{\]}\|'<,.>?\/\";:£§º©®\s]+[.｡。･・︒។։။۔።।][^0-9`~!@#$%^&*()+\-_=[{\]}\|'<,.>?\/\";:£§º©®\s]+/";
+    preg_match_all($regex, $input, $matches);
+    if (count($matches) === 1) {
+      return $matches[0];
+    }
+    return array();
+  }
+
+  // Determines of the string passed in is a real three word address.
```
Comment (imndaiga) [inline src/Geocoder.php]:
```suggestion
  // Determines if the string passed in is a real three word address.
```

File: src/Geocoder.php [hunk 3]
```diff
@@ -8,216 +8,252 @@
...
+    curl_close($handle);
+
+    // return the json
+    return $output;
+  }
+
+}
+
+
+// These are static funcntions that make options to be passed into autosuggest
```
Comment (imndaiga) [inline src/Geocoder.php]:
```suggestion
// These are static functions that return options to be passed into autosuggest
```

File: src/Geocoder.php [hunk 4]
```diff
@@ -8,216 +8,252 @@
...
-        // return the json
-        return $output;
-        }
+class Geocoder
+{
+
+  private $version = "3.4.0";  // if changing this, remember to change the comment block at the top, and match everything with the git tag
+  private $apiKey = "";
+  private $error = false;
+  private $baseUrl = 'https://api.what3words.com/v3/';
```
Comment (imndaiga) [inline src/Geocoder.php:24]: I reckon the camelCase prop name changes would technically make this a breaking change? i.e. `version=4.0.0`

**Review by imndaiga — COMMENTED** [review]
Main comment for thought would be the name prop changes re: avoiding breaking changes

File: src/Geocoder.php (hunk 3 above)
Comment (chaddgrimm) [reply src/Geocoder.php]: Technically this still creates/make options but I guess the message still not quite relayed correctly.

File: src/Geocoder.php (hunk 4 above)
Comment (chaddgrimm) [reply src/Geocoder.php:24]: these prop names were only ever used privately (or at least it should), specifically with apiKey and baseUrl. I'm unsure if there's a use case wherein anyone would use `$geocoder->baseUrl` or `$geocoder->apiKey` from outside the Geocoder class. Unless we're exposing these properties publicly, I doubt we would for api key.

Comment (imndaiga) [reply src/Geocoder.php:24]: yeah my concern was that this is a public package so we don't really know what it's downstream usage looks like. Also not sure how php handles package dep management i.e. are versions pinned by default. Happy to approve, good enough to sign post and keep an eye post release on bd queries

**Review by imndaiga — APPROVED** [review]

File: src/Geocoder.php (hunk 4 above)
Comment (chaddgrimm) [reply src/Geocoder.php:24]: Sounds good. Yeah, a very valid concern, this is to be published to packagist from what I know, how the package dep works, is also beyond me.

**Pull request merged**